torchvision/datasets/standford_cars.py

```diff
@@ -0,0 +1,129 @@
+import os
+import os.path
+from typing import Callable, Optional
+
+import scipy.io as io
+from PIL import Image
+
+from .utils import download_and_extract_archive, verify_str_arg, download_url
+from .vision import VisionDataset
+
+
+class StanfordCars(VisionDataset):
+    """`Stanford Cars <https://ai.stanford.edu/~jkrause/cars/car_dataset.html>`_ Dataset
+
+    .. warning::
+
+        This class needs `scipy <https://docs.scipy.org/doc/>`_ to load target files from `.mat` format.
+
+    Args:
+        root (string): Root directory of dataset
+        train (bool, optional):
+        download (bool, optional): If True, downloads the dataset from the internet and
+            puts it in root directory. If dataset is already downloaded, it is not
+            downloaded again.
+        transform (callable, optional): A function/transform that  takes in an PIL image
+            and returns a transformed version. E.g, ``transforms.RandomCrop``
+        target_transform (callable, optional): A function/transform that takes in the
+            target and transforms it."""
+
+    urls = (
+        "https://ai.stanford.edu/~jkrause/car196/cars_test.tgz",
+        "https://ai.stanford.edu/~jkrause/car196/cars_train.tgz",
+    )
+
+    md5s = ("4ce7ebf6a94d07f1952d94dd34c4d501", "065e5b463ae28d29e77c1b4b166cfe61")
+
+    annot_urls = (
+        "https://ai.stanford.edu/~jkrause/car196/cars_test_annos_withlabels.mat",
+        "https://ai.stanford.edu/~jkrause/cars/car_devkit.tgz",
+    )
+    extension = ".jpg"
+
+    def __init__(
+        self,
+        root: str,
+        train: bool = True,
+        transform: Optional[Callable] = None,
+        target_transform: Optional[Callable] = None,
+        download: bool = False,
+    ) -> None:
+        super().__init__(root, transform=transform, target_transform=target_transform)
+        self._download = download
+        self.root = os.path.expanduser(root)
+        self.train = train
+        self.url = self.urls[self.train]
+
+        if self._download:
+            self.download()
+
+        if not self._check_exists():
+            raise RuntimeError("Dataset not found. You can use download=True to download it")
+
+        self.class_names = self._get_class_names()
+        self.annotations = self._get_annotations()
+
+    def _get_class_names(self) -> dict:
+        """
+        Returns Mapping of class ids to class names in form of Dictionary
+        """
+        meta_data = io.loadmat(os.path.join(self.root, "devkit/cars_meta.mat"))
+        class_names = meta_data["class_names"][0]
+        return {
+            # Format class names appropriately for directory creation.
+            class_name[0].replace(" ", "_").replace("/", "_"): i
+            for i, class_name in enumerate(class_names)
+        }
+
+    def _get_annotations(self):
```

Comment:
These are not just annotations, but rather annotations and image file names, right? Thus, I think something like 

```suggestion
    def _make_dataset(self):
```

Would be more fitting.

Reply:
Plus, I would move all processing that is done in `__getitem__` here, e.g. let this create a `self._samples` that can be accessed like this:

```python
image_file, target = self._samples[idx]
image = Image.open(image_path).convert("RGB")

if transform:
    ...

...
```